pkg/storage/stores/local/archiver.go

```diff
@@ -0,0 +1,493 @@
+package local
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/aws"
+	"github.com/cortexproject/cortex/pkg/chunk/azure"
+	"github.com/cortexproject/cortex/pkg/chunk/gcp"
+	"github.com/cortexproject/cortex/pkg/chunk/local"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"go.etcd.io/bbolt"
+)
+
+const (
+	// ArchiverModeReadWrite is to allow both read and write
+	ArchiverModeReadWrite = iota
+	// ArchiverModeReadOnly is to allow only read operations
+	ArchiverModeReadOnly
+	// ArchiverModeWriteOnly is to allow only write operations
+	ArchiverModeWriteOnly
+
+	ArchiverFileUploadInterval = 15 * time.Minute
+	CacheCleanupInterval       = 24 * time.Hour
+)
+
+type StoreConfig struct {
+	Store            string                  `yaml:"store"`
+	AWSStorageConfig aws.StorageConfig       `yaml:"aws"`
+	GCSConfig        gcp.GCSConfig           `yaml:"gcs"`
+	FSConfig         local.FSConfig          `yaml:"filesystem"`
+	Azure            azure.BlobStorageConfig `yaml:"azure"`
+}
+
+type ArchiverConfig struct {
+	Enable          bool          `yaml:"enable"`
+	CacheLocation   string        `yaml:"cache_location"`
+	CacheTTL        time.Duration `yaml:"cache_ttl"`
+	StoreConfig     StoreConfig   `yaml:"store_config"`
+	ResyncInterval  time.Duration `yaml:"resync_interval"`
+	IngesterName    string        `yaml:"-"`
+	Mode            int           `yaml:"-"`
+	BoltdbDirectory string        `yaml:"-"`
+}
+
+// RegisterFlags registers flags.
+func (cfg *ArchiverConfig) RegisterFlags(f *flag.FlagSet) {
+	storeFlagsPrefix := "boltdb.archiver."
+	cfg.StoreConfig.AWSStorageConfig.RegisterFlagsWithPrefix(storeFlagsPrefix, f)
+	cfg.StoreConfig.GCSConfig.RegisterFlagsWithPrefix(storeFlagsPrefix, f)
+	cfg.StoreConfig.FSConfig.RegisterFlagsWithPrefix(storeFlagsPrefix, f)
+
+	f.BoolVar(&cfg.Enable, "boltdb.archiver.enable", false, "Enable archival of boltdb files to a store")
+	f.StringVar(&cfg.StoreConfig.Store, "boltdb.archiver.store", "filesystem", "Store for keeping boltdb files")
+	f.StringVar(&cfg.CacheLocation, "boltdb.archiver.cache-location", "", "Cache location for restoring boltDB files for queries")
+	f.DurationVar(&cfg.CacheTTL, "boltdb.archiver.cache-ttl", 24*time.Hour, "TTL for boltDB files restored in cache for queries")
+	f.DurationVar(&cfg.ResyncInterval, "boltdb.archiver.resync-interval", 5*time.Minute, "Resync downloaded files with the store")
```

Comment:
Your only documentation is those flags and I think they need a bit more details. Unless you plan to add a full page of documentation.

Reply:
I think I will have to add full page documentation about running it because we can't put too much in help text of flags.

Reply:
I think it's a good call !